common/timing.cc

```diff
@@ -36,115 +37,263 @@ struct Timing
+                    // Also add I/O cells too
+                                                      // TODO(eddieh): More generic way of detecting PLLs
+                    if (is_io || cell.second->type == ctx->id("ICESTORM_PLL")) {
+                        topographical_order.emplace_back(o->net);
+                        net_data.emplace(o->net, TimingData{});
+                    }
+                    // Otherwise, for all driven input ports on this cell,
+                    //   if a timing arch exists between the input and
+                    //   the current output port, increment fanin counter
+                    for (auto i : input_ports) {
+                        DelayInfo comb_delay;
+                        bool is_path = ctx->getCellDelay(cell.second.get(), i, o->name, comb_delay);
+                        if (is_path)
+                            port_fanin[o]++;
+                    }
+                }
             }
-            if (crit_path)
-                current_path.pop_back();
         }
-        return net_budget;
-    }
 
-    // Follow a path, returning budget to annotate
-    delay_t follow_user_port(PortRef &user, int path_length, delay_t slack)
-    {
-        delay_t value;
-        if (ctx->getPortClock(user.cell, user.port) != IdString()) {
-            // At the end of a timing path (arguably, should check setup time
-            // here too)
-            value = slack / path_length;
-            if (slack < min_slack) {
-                min_slack = slack;
-                if (crit_path)
-                    *crit_path = current_path;
+        // If these constant nets exist, add them to the topographical ordering too
+        auto it = ctx->nets.find(ctx->id("$PACKER_VCC_NET"));
+        if (it != ctx->nets.end()) {
+            topographical_order.emplace_back(it->second.get());
+            net_data.emplace(it->second.get(), TimingData{});
+        }
+        it = ctx->nets.find(ctx->id("$PACKER_GND_NET"));
+        if (it != ctx->nets.end()) {
+            topographical_order.emplace_back(it->second.get());
+            net_data.emplace(it->second.get(), TimingData{});
+        }
+
+        std::deque<NetInfo*> queue(topographical_order.begin(), topographical_order.end());
+
+        // Now walk the design, from the start points identified previously, building
+        //   up a topographical order
+        while (!queue.empty()) {
+            const auto net = queue.front();
+            queue.pop_front();
+
+            DelayInfo clkToQ;
+            for (auto &usr : net->users) {
+                auto clock_domain = ctx->getPortClock(usr.cell, usr.port);
+                for (auto& port : usr.cell->ports) {
+                    if (port.second.type == PORT_OUT && port.second.net) {
+                        // Skip if this is a clocked output (but allow non-clocked ones)
+                        if (clock_domain != IdString() && ctx->getCellDelay(usr.cell, clock_domain, port.first, clkToQ))
```

Comment:
@daveshah1 This _should_ allow cells which have both clocked and unclocked outputs to be analysed which might fulfill #32 --- this was necessary because I found a case where an I3 -> COUT path wasn't being analysed since the FF was also present.